bigtable/bttest/grpc_error_injection.go

```diff
@@ -0,0 +1,315 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+This file provides interceptors to hook into grpc streaming calls
+and inject latency and errors into cbtemulator for testing
+*/
+package bttest
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var (
+	validStreamMethodSuffixes = []string{"ReadRows", "MutateRows"}
+)
+
+/*
+   Creates interceptors inject latency or errors into cbtemulator
+*/
+type EmulatorInterceptorBuilder struct {
+	LatencyTargets       latencyTargets
+	GrpcErrorCodeTargets grpcErrorCodeTargets
+}
+
+func (esib *EmulatorInterceptorBuilder) BuildStreamInterceptor() grpc.ServerOption {
+	return grpc.StreamInterceptor(esib.interceptorFunc())
+}
+
+/*
+   Stack error targets are so that arguments like:
+     - [ReadRows:20%:12, ReadRows:10%:14]
+   transform to:
+     - [ReadRows:20%:12, ReadRows:30%:14]
+   Thus, when we check against randint(100), we'll
+     throw the first error 20% of the time and the second error 10% of the time
+*/
+func (esib *EmulatorInterceptorBuilder) stackGrpcErrorCodeTargets() {
+	sort.SliceStable(esib.GrpcErrorCodeTargets, func(i, j int) bool {
```

Comment:
The side effect of mutating the underlying data feels a bit hacky here and will have really bad consequences if someone calls interceptorFunc() multiple times (and I wont blame them because it looks like a getter :) )

Reply:
Also I dont quite understand the need for stacking them

Reply:
Good call - moved the stacking logic into the `Set()` function.